File: rust/glue-core/src/router.rs
// glue-core/src/router.rs — capability-based agent selection

use crate::capabilities::CapabilityMatch;

pub struct RouteResult {
    pub agent_id: String,
    pub protocol: String,
    pub endpoint: String,
    pub reason: String,
}

pub struct Router {
    min_health: Option<String>,
    min_trust_score: Option<f64>,
}

impl Router {
    pub fn new() -> Self {
        Self {
            min_health: None,
            min_trust_score: None,
        }
    }

    pub fn with_min_health(mut self, health: &str) -> Self {
        self.min_health = Some(health.to_string());
        self
    }

    pub fn with_min_trust(mut self, score: f64) -> Self {
        self.min_trust_score = Some(score);
        self
    }

    pub fn select(&self, candidates: &[CapabilityMatch]) -> Result<RouteResult, String> {
        if candidates.is_empty() {
            return Err("No candidates available for routing".to_string());
        }

        // Filter by health
        let mut filtered: Vec<&CapabilityMatch> = candidates.iter().filter(|c| {
            match self.min_health.as_deref() {
                Some("healthy") => c.health == "healthy",
                Some("degraded") => c.health == "healthy" || c.health == "degraded",
                _ => c.health != "unreachable",
            }
        }).collect();

        if filtered.is_empty() {
            filtered = candidates.iter().collect();
        }

        // Filter by trust score
        if let Some(min_trust) = self.min_trust_score {
            let trust_filtered: Vec<&CapabilityMatch> = filtered.iter()
                .filter(|c| c.trust_score >= min_trust)
                .cloned()
                .collect();
            if !trust_filtered.is_empty() {
                filtered = trust_filtered;
            }
        }

        // Score: prefer higher trust, then healthier agents
        let mut scored: Vec<(f64, &CapabilityMatch)> = filtered.iter().map(|c| {
            let health_score = if c.health == "healthy" { 0.4 } else { 0.0 };
            let score = c.trust_score * 0.6 + health_score;
            (score, *c)
        }).collect();

        scored.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));

        let selected = scored[0].1;
        Ok(RouteResult {
            agent_id: selected.agent_id.clone(),
            protocol: selected.protocol.clone(),
            endpoint: selected.endpoint.clone(),
            reason: format!(
                "Selected by capability match (trust: {:.2}, health: {})",
                selected.trust_score, selected.health
            ),
        })
    }
}

File: packages/core/src/router/router.ts
// Router — capability-based agent selection

import type { CapabilityId } from '@glue/contracts';
import type { AgentRecord } from '@glue/contracts';
import type { CapabilityMatch } from '../capabilities/capability-graph.js';

export interface RouteResult {
  agentId: string;
  protocol: string;
  endpoint: string;
  reason: string;
}

export interface RouterOptions {
  /** Require a minimum health status */
  minHealth?: 'degraded' | 'healthy';
  /** Require a minimum trust score (0-1) */
  minTrustScore?: number;
}

export class Router {
  constructor(private options: RouterOptions = {}) {}

  async select(candidates: CapabilityMatch[]): Promise<RouteResult> {
    if (candidates.length === 0) {
      throw new Error('No candidates available for routing');
    }

    // Filter by health
    let filtered = candidates.filter(c => {
      if (this.options.minHealth === 'healthy') return c.health === 'healthy';
      if (this.options.minHealth === 'degraded') return c.health === 'healthy' || c.health === 'degraded';
      return c.health !== 'unreachable';
    });

    if (filtered.length === 0) {
      // Fallback to any candidate if all filtered out
      filtered = candidates;
    }

    // Filter by trust score
    if (this.options.minTrustScore !== undefined) {
      filtered = filtered.filter(c => (c.trustScore ?? 0) >= this.options.minTrustScore!);
      if (filtered.length === 0) filtered = candidates;
    }

    // Score: prefer higher trust, then healthier agents
    const scored = filtered.map(c => ({
      ...c,
      _score: c.trustScore * 0.6 + (c.health === 'healthy' ? 0.4 : 0),
    }));

    scored.sort((a, b) => b._score - a._score);

    const selected = scored[0];
    return {
      agentId: selected.agentId,
      protocol: selected.protocol,
      endpoint: selected.endpoint,
      reason: `Selected by capability match (trust: ${selected.trustScore.toFixed(2)}, health: ${selected.health})`,
    };
  }
}
